v2/spiffeid/id.go

```diff
@@ -1,78 +1,162 @@
 package spiffeid
 
 import (
+	"errors"
+	"fmt"
 	"net/url"
+	"strings"
 )
 
 // ID is a SPIFFE ID
 type ID struct {
+	td   TrustDomain
+	path string
 }
 
 // New creates a new ID using the trust domain (e.g. example.org) and path
 // segments. An error is returned if the trust domain is not valid (see
 // TrustDomainFromString).
+// Warning: Percent encoded characters on path segments are
+// not decoded, instead this function will percent encode symbols
+// when the ID is converted to a string.
 func New(trustDomain string, segments ...string) (ID, error) {
-	panic("not implemented")
+	td, err := TrustDomainFromString(trustDomain)
+	if err != nil {
+		return ID{}, err
+	}
+
+	path := strings.Join(segments, "/")
+	if len(path) > 0 {
+		path = "/" + path
+	}
+
+	return ID{
+		td:   td,
+		path: path,
+	}, nil
 }
 
 // Must creates a new ID using the trust domain (e.g. example.org) and path
 // segments. The function panics if the trust domain is not valid (see
 // TrustDomainFromString).
 func Must(trustDomain string, segments ...string) ID {
-	panic("not implemented")
+	id, err := New(trustDomain, segments...)
+	if err != nil {
+		panic(err)
+	}
+	return id
 }
 
 // Join returns the string representation of an ID inside the given trust
 // domain (e.g. example.org) with the given path segments. An error is returned
 // if the trust domain is not valid (see TrustDomainFromString).
 func Join(trustDomain string, segments ...string) (string, error) {
-	panic("not implemented")
+	id, err := New(trustDomain, segments...)
+	if err != nil {
+		return "", err
+	}
+
+	return id.String(), nil
 }
 
 // MustJoin returns the string representation of an ID inside the given trust
 // domain (e.g. example.org) with the given path segments. The function panics
 // if the trust domain is not valid (see TrustDomainFromString).
 func MustJoin(trustDomain string, segments ...string) string {
-	panic("not implemented")
+	idstr, err := Join(trustDomain, segments...)
+	if err != nil {
+		panic(err)
+	}
+	return idstr
 }
 
 // FromString parses a SPIFFE ID from a string.
 func FromString(s string) (ID, error) {
-	panic("not implemented")
+	uri, err := url.Parse(s)
+	if err != nil {
+		return ID{}, fmt.Errorf("invalid SPIFFE ID: %v", err)
+	}
+
+	return FromURI(uri)
 }
 
 // FromURI parses a SPIFFE ID from a URI.
-func FromURI(u *url.URL) (ID, error) {
-	panic("not implemented")
+func FromURI(uri *url.URL) (ID, error) {
+	if uri == nil || *uri == (url.URL{}) {
+		return ID{}, errors.New("invalid SPIFFE ID: SPIFFE ID is empty")
+	}
+
+	// General validation
+	switch {
+	case strings.ToLower(uri.Scheme) != "spiffe":
+		return ID{}, errors.New("invalid SPIFFE ID: invalid scheme")
+	case uri.User != nil:
+		return ID{}, errors.New("invalid SPIFFE ID: user info is not allowed")
+	case uri.Host == "":
+		return ID{}, errors.New("invalid SPIFFE ID: trust domain is empty")
+	case uri.Port() != "":
```

Comment:
There is a missing condition here. Even if `Port()` is empty, the SPIFFE ID should still have the `:` delimiter, which is against the SPIFFE ID spec (e.g. `"spiffe://example.org:/foo/bar"`). We should add a check that the `Host` doesn't have a trailing `:` after the check for an empty port.